File: TeamCode/src/main/java/org/firstinspires/ftc/teamcode/classes/Pose2d.java
package org.firstinspires.ftc.teamcode.classes;

public class Pose2d {
    public double x;
    public double y;
    public double theta;

    public Pose2d (double x, double y, double theta) {
        this.x = x;
        this.y = y;
        this.theta = theta;
    }

    public double getX() {
        return x;
    }

    public double getY() {
        return y;
    }

    public double getTheta() {
        return theta;
    }

    public Vector2d headingVec() {
        return new Vector2d(Math.cos(theta), Math.sin(theta));
    }

    public Pose2d times(double scalar) {
        return new Pose2d(scalar * x, scalar * y, scalar * theta);
    }
}

File: RoadRunner/src/main/kotlin/com/acmerobotics/roadrunner/geometry/Pose2d.kt
package com.acmerobotics.roadrunner.geometry

import com.acmerobotics.roadrunner.util.Angle
import com.acmerobotics.roadrunner.util.epsilonEquals
import kotlin.math.cos
import kotlin.math.sin

/**
 * Class for representing 2D robot poses (x, y, and heading) and their derivatives.
 */
data class Pose2d @JvmOverloads constructor(
    val x: Double = 0.0,
    val y: Double = 0.0,
    val heading: Double = 0.0
) {
    constructor(pos: Vector2d, heading: Double) : this(pos.x, pos.y, heading)

    fun vec() = Vector2d(x, y)

    fun headingVec() = Vector2d(cos(heading), sin(heading))

    operator fun plus(other: Pose2d) =
        Pose2d(x + other.x, y + other.y, heading + other.heading)

    operator fun minus(other: Pose2d) =
        Pose2d(x - other.x, y - other.y, heading - other.heading)

    operator fun times(scalar: Double) =
        Pose2d(scalar * x, scalar * y, scalar * heading)

    operator fun div(scalar: Double) =
        Pose2d(x / scalar, y / scalar, heading / scalar)

    operator fun unaryMinus() = Pose2d(-x, -y, -heading)

    infix fun epsilonEquals(other: Pose2d) =
        x epsilonEquals other.x && y epsilonEquals other.y && heading epsilonEquals other.heading

    infix fun epsilonEqualsHeading(other: Pose2d) =
        x epsilonEquals other.x && y epsilonEquals other.y && Angle.normDelta(heading - other.heading) epsilonEquals 0.0

    override fun toString() = String.format("(%.3f, %.3f, %.3f°)", x, y, Math.toDegrees(heading))
}

operator fun Double.times(pose: Pose2d) = pose.times(this)

operator fun Double.div(pose: Pose2d) = pose.div(this)
